runtime/src/iree/hal/drivers/webgpu/platform/emscripten/emscripten_driver.c

```diff
@@ -0,0 +1,249 @@
+//===----------------------------------------------------------------------===//
+
+IREE_API_EXPORT void iree_hal_webgpu_driver_options_initialize(
+    iree_hal_webgpu_driver_options_t* out_options) {
+  IREE_ASSERT_ARGUMENT(out_options);
+  memset(out_options, 0, sizeof(*out_options));
+
+  out_options->backend_preference = IREE_HAL_WEBGPU_DRIVER_BACKEND_ANY;
+  out_options->log_level = IREE_HAL_WEBGPU_DRIVER_LOG_LEVEL_OFF;
+
+  // TODO(benvanik): coming in future spec update. For now go high-perf.
+  // out_options->power_preference = WGPUPowerPreference_Undefined;
+  out_options->power_preference = WGPUPowerPreference_HighPerformance;
+
+  iree_hal_webgpu_device_options_initialize(&out_options->device_options);
+}
+
+//===----------------------------------------------------------------------===//
+// Synchronous adapter and device request functions
+//===----------------------------------------------------------------------===//
+
+// We use Asyncify for convenience - for now.
+//   https://emscripten.org/docs/porting/asyncify.html
+//   https://github.com/emscripten-core/emscripten/issues/15746
+//   https://github.com/juj/wasm_webgpu/blob/main/lib/lib_webgpu.h
+//
+// The requestAdapter and requestDevice functions are asynchronous, while the
+// HAL API has synchronous driver and device creation. We want to support both
+// platform-independent tests (CTS tests, 'check' framework tests) and user
+// applications. For platform-independent tests, we keep the APIs synchronous
+// by using Asyncify. For user applications, we could pass in a loop or add
+// asynchronous APIs that let those applications use async/await directly.
+//
+// An even simpler solution for user applications is to request an adapter and
+// and device purely up in JavaScript, then to pass the already created device
+// in via preinitializedWebGPUDevice / emscripten_webgpu_get_device().
+
+#ifdef EM_ASYNC_JS
+
+EM_ASYNC_JS(WGPUAdapter, wgpuInstanceRequestAdapterSync, (), {
+  // TODO(scotttodd): WGPURequestAdapterOptions struct
+  const adapter = await navigator['gpu']['requestAdapter']();
+  // WARNING: this calls functions directly on Emscripten's library_webgpu.js.
+  // This is not a stable API!
+  const adapterId = WebGPU.mgrAdapter.create(adapter);
+  return adapterId;
+});
+
+EM_ASYNC_JS(WGPUDevice, wgpuAdapterRequestDeviceSync, (WGPUAdapter adapterId), {
+  // WARNING: this calls functions directly on Emscripten's library_webgpu.js.
+  // This is not a stable API!
+  const adapter = WebGPU.mgrAdapter.get(adapterId);
+
+  // TODO(scotttodd): WGPUDeviceDescriptor struct
+  const descriptor = {};
+  const device = await adapter['requestDevice'](descriptor);
+
+  const deviceWrapper = {queueId : WebGPU.mgrQueue.create(device["queue"])};
+  const deviceId = WebGPU.mgrDevice.create(device, deviceWrapper);
+  return deviceId;
+});
+
+#else
+
+WGPUAdapter wgpuInstanceRequestAdapterSync() {
+  fprintf(stderr, "wgpuInstanceRequestAdapterSync requires -sASYNCIFY\n");
```

Comment:
is this the only place (so far) that needs asyncify? could have a global device that we query here (as mentioned in the comment above) but probably easiest to wait until we have everything working and then try to nuke asyncify altogether

Reply:
Yep. Also worth calling out is "mapSync", which I couldn't find a browser-friendly solution for (didn't look that deeply though)

```c++
IREEWGPUBufferMapSyncStatus iree_wgpuBufferMapSync(WGPUDevice device,
                                                   WGPUBuffer buffer,
                                                   WGPUMapModeFlags mode,
                                                   size_t offset, size_t size) {
  IREEWGPUBufferMapSyncStatus status = IREEWGPUBufferMapSyncStatus_Unknown;
  wgpuBufferMapAsync(buffer, mode, offset, size,
                     iree_hal_webgpu_buffer_map_sync_callback, &status);
  // TODO(scotttodd): poll / wait somehow, or implement sync mapping differently
  // wgpuDevicePoll(device, /*force_wait=*/true);
  return status;
}
```

Reply:
Yeah that'll be the big one - until we can remove that we'll probably be stuck with asyncify ;(